### ISSUE TITLE
Creating Package Extensions for Language Server?

### ISSUE THREAD

**Issue by nateybear** [julia-vscode/LanguageServer.jl#1309]
I'm working on a package extension to hook into the language server's `StaticLint.get_path` function and patch it by resolving project-relative paths from `DrWatson`. This may be my lack of understanding about package extensions and `LanguageServer` combining here, but I can't get the package extension to actually load inside of the language server process. Some notes:

1. If I add `LanguageServer` as an explicit dependency in my project, then I can get the extension to load in my Julia REPL.
2. I assume the language server processes in VS Code, nvim, etc. start up in a different environment than my own Julia environment. How does this work, i.e. if I don't specify an explicit dependency in my project, will this still trigger the package extension to load?
3. Is `LanguageServer` actually *loading* the packages in my project? I'm guessing it's just inspecting them for symbols. I'm not sure when the extension code is actually executed, but my guess would be when the package is loaded. So is there no way for me to get the extension code to execute inside the language server process in this case?

Any guidance on how to approach this is much appreciated!

**Comment by davidanthoff** [comment]
The LS process never loads any user packages or user code and there is no mechanism to extend the LS with custom code inside the LS process. That design is very much on purpose, we want to make sure that errors in packages ideally can't take the LS process down.

We do sometimes spawn an additional child process where we load user packages to extract symbols etc, but that is then a special purpose process that doesn't run the regular LS stuff, but really only extracts the symbols and then shuts down again.

Can you give an example of the kind of `include` that you are trying to enable, that currently doesn't work? If it is generic enough, maybe we can just generally try to make it work in the LS.

**Comment by nateybear** [comment]
`DrWatson` has project-relative path functions, like `here` in R or `pyprojroot` in Python. The definitions are essentially

```julia
projectdir(x...) = joinpath(Base.active_project(), x...)
srcdir(x...) = joinpath(Base.active_project(), "src", x...)
# etc...
```

And the includes would look like `include(srcdir("somefunction.jl"))`. The POC I have just checks if a `CSTParser` expression looks like a call of one of these functions, and if so it reconstructs the original code string and evals it.

In terms of being generic enough, it's specific to a package and I recall some discussion in the issues here about not wanting to try and accommodate too many packages inside of `LanguageServer.jl` itself. OTOH, I would argue that project-relative paths are an important programming idiom that's well supported in other languages, and to support it here we have to choose a package that provides them and accommodate it.

**Comment by nateybear** [comment]
@davidanthoff just nudging for possible follow-up

**Comment by davidanthoff** [comment]
So my gut feeling is that this is too specific for the LS... I think specifically for `include` I also think just regular relative paths seem fine? Could that be a solution for Dr. Watson projects, to use regular relative paths for `include`, and then these functions for things like access to data etc?

**Comment by nateybear** [comment]
@davidanthoff Thanks for your response. I'm fine with the decision that it's too specific for the language server. As an alternative I wish there were some way to hook into the lsp to alter its behavior with userland code. (Another pain point I have is not expanding certain macros like chain, lazy's pipes, or macros that define new variables. I know the argument about macros being able to execute arbitrary code and the dangers therein, but having the users opt into the footgun is less bad imo.) I'll close the issue for now as it's no longer relevant.

**Comment by davidanthoff** [comment]
We are reconsidering how we are treating macros, so that might someday properly work. For that we would probably have to spin up another Julia process where we load all the user code, and once that is in place that might also be an opportunity to add other hooks to the LS as we could then make sure nothing can crash the LS. But realistically, all of that is quite far away...

**Comment by nateybear** [comment]
@davidanthoff great to know. If there is scope for contributions on that I'm happy to contribute.
